nexus/types/src/deployment/blueprint_diff.rs

```diff
@@ -11,26 +11,621 @@ use super::blueprint_display::{
+    ) -> Option<&'a BlueprintDatasetsConfig> {
+        self.diff.blueprint_datasets.added.get(sled_id).cloned()
+    }
+
+    /// Return the `BlueprintDatasetsConfig` for a removed sled
+    pub fn datasets_on_removed_sled(
+        &self,
+        sled_id: &SledUuid,
+    ) -> Option<&'a BlueprintDatasetsConfig> {
+        self.diff.blueprint_datasets.removed.get(sled_id).cloned()
+    }
+
+    /// Return the `BlueprintDatasetsConfigDiff` for a modified sled
+    pub fn datasets_on_modified_sled(
+        &self,
+        sled_id: &SledUuid,
+    ) -> Option<&'a BlueprintDatasetsConfigDiff> {
+        self.diff.blueprint_datasets.modified.get(sled_id)
+    }
+
+    /// Iterate over all added zones on a sled
+    pub fn added_zones(&self, sled_id: &SledUuid) -> Option<BpDiffZoneDetails> {
+        // First check if the sled is added
+        if let Some(&zones_cfg) = self.diff.blueprint_zones.added.get(sled_id) {
+            return Some(BpDiffZoneDetails::new(
+                None,
+                Some(zones_cfg.generation),
+                zones_cfg.zones.iter(),
+            ));
+        }
+
+        // Then check if the sled is modified and there are any added zones
+        self.diff.blueprint_zones.modified.get(sled_id).map(|zones_cfg_diff| {
+            BpDiffZoneDetails::new(
+                Some(*zones_cfg_diff.generation.before),
+                Some(*zones_cfg_diff.generation.after),
+                zones_cfg_diff.zones.added.values().map(|z| *z),
+            )
+        })
+    }
+
+    /// Iterate over all removed zones on a sled
+    pub fn removed_zones(
+        &self,
+        sled_id: &SledUuid,
+    ) -> Option<BpDiffZoneDetails> {
+        // First check if the sled is removed
+        if let Some(&zones_cfg) = self.diff.blueprint_zones.removed.get(sled_id)
+        {
+            return Some(BpDiffZoneDetails::new(
+                Some(zones_cfg.generation),
+                None,
+                zones_cfg.zones.iter(),
+            ));
+        }
+
+        // Then check if the sled is modified and there are any removed zones
+        self.diff.blueprint_zones.modified.get(sled_id).map(|zones_cfg_diff| {
+            BpDiffZoneDetails::new(
+                Some(*zones_cfg_diff.generation.before),
+                Some(*zones_cfg_diff.generation.after),
+                zones_cfg_diff.zones.removed.values().map(|z| *z),
+            )
+        })
+    }
+
+    /// Iterate over all modified zones on a sled
+    pub fn modified_zones(
+        &'a self,
+        sled_id: &SledUuid,
+    ) -> Option<(BpDiffZonesModified, BpDiffZoneErrors)> {
+        // Then check if the sled is modified and there are any added zones
+        self.diff.blueprint_zones.modified.get(sled_id).map(|zones_cfg_diff| {
+            BpDiffZonesModified::new(
+                *zones_cfg_diff.generation.before,
+                *zones_cfg_diff.generation.after,
+                zones_cfg_diff.zones.modified.values(),
```

Comment:
Could the sled have been modified but no _zones_ modified? Should we return `None` in that case? (I guess the same question applies to all the other methods that have this pattern.)

Reply:
Hmm. Yes, you are right. I think we should return `None` in these cases.

Reply:
Done in [69fe736](https://github.com/oxidecomputer/omicron/pull/7466/commits/69fe736d8238c3b75d152f73d053ec9b15c6e7da)